File: My_Library/app/static/js/project.js
/* === 작업실 탭 === */

const btnNewProject = document.getElementById("btnNewProject");

btnNewProject.addEventListener("click", () => {
  const name = prompt("새 프로젝트 이름을 입력하세요:");
  if (!name) return;
  const desc = prompt("프로젝트 설명 (선택):") || "";
  createProject(name, desc);
});

async function createProject(name, description) {
  try {
    const res = await api("/api/projects/", {
      method: "POST",
      body: { name, description },
    });
    showToast(`프로젝트 "${name}" 생성 완료!`, "success");
    loadProjects();
  } catch (err) {
    showToast(`생성 실패: ${err.message}`, "error");
  }
}

async function loadProjects() {
  try {
    const res = await api("/api/projects/");
    const projects = res.data;
    const container = document.getElementById("projectList");
    const detail = document.getElementById("projectDetail");
    detail.style.display = "none";

    if (projects.length === 0) {
      container.innerHTML = '<p class="placeholder-text">프로젝트가 없습니다. "새 프로젝트" 버튼을 클릭해보세요.</p>';
      return;
    }

    container.innerHTML = projects
      .map((p) => {
        const statusClass =
          p.status === "진행중" ? "ongoing" : p.status === "완료" ? "done" : "paused";
        return `
        <div class="project-card" onclick="showProjectDetail(${p.id})">
          <div style="display:flex;justify-content:space-between;align-items:center;">
            <div class="project-name">${p.name}</div>
            <span class="status-badge ${statusClass}">${p.status}</span>
          </div>
          <div class="project-meta">
            ${p.description ? `<span>${p.description}</span> · ` : ""}
            <span>자료 ${p.material_count}개</span>
          </div>
        </div>`;
      })
      .join("");
  } catch (err) {
    showToast(`프로젝트 로드 실패: ${err.message}`, "error");
  }
}

async function showProjectDetail(projectId) {
  try {
    const res = await api(`/api/projects/${projectId}`);
    const p = res.data;
    const detail = document.getElementById("projectDetail");

    const materials = (p.materials || [])
      .map(
        (m) => `
      <div class="card" style="padding:12px;">
        <div style="display:flex;justify-content:space-between;align-items:center;">
          <div>
            <strong>${m.title}</strong>
            <div style="font-size:0.8rem;color:var(--text-muted);">${m.category}${m.note ? " · " + m.note : ""}</div>
          </div>
          <button class="btn btn-secondary" style="padding:4px 10px;font-size:0.78rem;"
            onclick="removeFromProject(${p.id}, ${m.id})">제거</button>
        </div>
      </div>`
      )
      .join("");

    const statusClass =
      p.status === "진행중" ? "ongoing" : p.status === "완료" ? "done" : "paused";

    detail.style.display = "block";
    detail.innerHTML = `
      <div class="settings-section" style="margin-top:20px;">
        <div style="display:flex;justify-content:space-between;align-items:center;margin-bottom:16px;">
          <h3>${p.name} <span class="status-badge ${statusClass}">${p.status}</span></h3>
          <div style="display:flex;gap:8px;">
            <button class="btn btn-secondary" onclick="addMaterialToProject(${p.id})">+ 자료 추가</button>
            <button class="btn btn-secondary" onclick="changeProjectStatus(${p.id})">상태 변경</button>
            <button class="btn btn-secondary" onclick="deleteProject(${p.id})">🗑️ 삭제</button>
          </div>
        </div>
        ${p.description ? `<p style="color:var(--text-secondary);margin-bottom:16px;">${p.description}</p>` : ""}
        <h4 style="margin-bottom:10px;">연결된 자료 (${p.material_count})</h4>
        <div class="card-list">${materials || '<p class="placeholder-text">연결된 자료가 없습니다.</p>'}</div>
      </div>`;
  } catch (err) {
    showToast(`프로젝트 상세 로드 실패: ${err.message}`, "error");
  }
}

let _addMatProjectId = null;

async function addMaterialToProject(projectId) {
  _addMatProjectId = projectId;
  const modal = document.getElementById("addMaterialModal");
  const listEl = document.getElementById("addMatList");
  const searchInput = document.getElementById("addMatSearchInput");
  searchInput.value = "";
  listEl.innerHTML = '<p class="placeholder-text">로딩 중...</p>';
  modal.style.display = "flex";
  await _loadAddMaterialList("");
}

async function _loadAddMaterialList(query) {
  const listEl = document.getElementById("addMatList");
  try {
    const params = new URLSearchParams({ page: "1", size: "50", status: "active" });
    if (query) params.set("q", query);
    const res = await api(`/api/library/materials?${params}`);
    const items = (res.data && res.data.items) || [];
    if (items.length === 0) {
      listEl.innerHTML = '<p class="placeholder-text">검색 결과가 없습니다.</p>';
      return;
    }
    listEl.innerHTML = items.map((m) => `
      <label class="add-mat-item" style="display:flex;align-items:center;gap:8px;padding:8px 10px;border-bottom:1px solid var(--border);cursor:pointer;">
        <input type="checkbox" value="${m.id}" class="add-mat-cb">
        <div style="flex:1;min-width:0;">
          <div style="font-weight:500;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;">${m.title}</div>
          <div style="font-size:0.78rem;color:var(--text-muted);">${m.category_large} > ${m.category_medium} · ID: ${m.id}</div>
        </div>
      </label>
    `).join("");
  } catch (err) {
    listEl.innerHTML = `<p class="placeholder-text">로드 실패: ${err.message}</p>`;
  }
}

function closeAddMaterialModal() {
  document.getElementById("addMaterialModal").style.display = "none";
  _addMatProjectId = null;
}

document.getElementById("btnAddMatSearch").addEventListener("click", () => {
  _loadAddMaterialList(document.getElementById("addMatSearchInput").value.trim());
});
document.getElementById("addMatSearchInput").addEventListener("keydown", (e) => {
  if (e.key === "Enter") _loadAddMaterialList(e.target.value.trim());
});

document.getElementById("btnAddMatConfirm").addEventListener("click", async () => {
  if (!_addMatProjectId) return;
  const checked = document.querySelectorAll("#addMatList .add-mat-cb:checked");
  if (checked.length === 0) {
    showToast("자료를 선택하세요.", "info");
    return;
  }
  let successCount = 0;
  for (const cb of checked) {
    try {
      await api(`/api/projects/${_addMatProjectId}/materials`, {
        method: "POST",
        body: { material_id: parseInt(cb.value) },
      });
      successCount++;
    } catch (err) {
      showToast(`ID ${cb.value} 추가 실패: ${err.message}`, "error");
    }
  }
  if (successCount > 0) {
    showToast(`${successCount}개 자료 추가 완료!`, "success");
    showProjectDetail(_addMatProjectId);
  }
  closeAddMaterialModal();
});

async function removeFromProject(projectId, materialId) {
  try {
    await api(`/api/projects/${projectId}/materials/${materialId}`, {
      method: "DELETE",
    });
    showToast("자료 제거 완료", "success");
    showProjectDetail(projectId);
  } catch (err) {
    showToast(`제거 실패: ${err.message}`, "error");
  }
}

async function changeProjectStatus(projectId) {
  const status = prompt("새 상태를 입력하세요 (진행중 / 완료 / 보류):");
  if (!status) return;

  try {
    await api(`/api/projects/${projectId}`, {
      method: "PUT",
      body: { status },
    });
    showToast("상태 변경 완료!", "success");
    showProjectDetail(projectId);
  } catch (err) {
    showToast(`변경 실패: ${err.message}`, "error");
  }
}

async function deleteProject(projectId) {
  if (!confirm("정말 삭제하시겠습니까?")) return;

  try {
    await api(`/api/projects/${projectId}`, { method: "DELETE" });
    showToast("프로젝트 삭제 완료", "success");
    loadProjects();
  } catch (err) {
    showToast(`삭제 실패: ${err.message}`, "error");
  }
}

File: My_Library/app/core/project.py
from sqlalchemy.orm import Session

from app.db.models import Project, ProjectMaterial, Material


def create_project(db: Session, name: str, description: str = "") -> dict:
    project = Project(name=name, description=description)
    db.add(project)
    db.commit()
    db.refresh(project)
    return _project_to_dict(project, db)


def list_projects(db: Session, status: str = "") -> list[dict]:
    q = db.query(Project)
    if status:
        q = q.filter(Project.status == status)
    projects = q.order_by(Project.updated_at.desc()).all()
    return [_project_to_dict(p, db) for p in projects]


def get_project_detail(db: Session, project_id: int) -> dict | None:
    project = db.query(Project).filter(Project.id == project_id).first()
    if not project:
        return None
    return _project_to_dict(project, db)


def update_project(
    db: Session,
    project_id: int,
    name: str | None = None,
    description: str | None = None,
    status: str | None = None,
) -> dict | None:
    project = db.query(Project).filter(Project.id == project_id).first()
    if not project:
        return None
    if name is not None:
        project.name = name
    if description is not None:
        project.description = description
    if status is not None:
        project.status = status
    db.commit()
    db.refresh(project)
    return _project_to_dict(project, db)


def delete_project(db: Session, project_id: int) -> bool:
    project = db.query(Project).filter(Project.id == project_id).first()
    if not project:
        return False
    db.delete(project)
    db.commit()
    return True


def add_material_to_project(
    db: Session,
    project_id: int,
    material_id: int,
    note: str = "",
) -> dict | None:
    project = db.query(Project).filter(Project.id == project_id).first()
    material = db.query(Material).filter(Material.id == material_id).first()
    if not project or not material:
        return None

    existing = (
        db.query(ProjectMaterial)
        .filter(
            ProjectMaterial.project_id == project_id,
            ProjectMaterial.material_id == material_id,
        )
        .first()
    )
    if existing:
        return _project_to_dict(project, db)

    link = ProjectMaterial(
        project_id=project_id,
        material_id=material_id,
        note=note,
    )
    db.add(link)
    db.commit()
    return _project_to_dict(project, db)


def remove_material_from_project(
    db: Session, project_id: int, material_id: int
) -> bool:
    link = (
        db.query(ProjectMaterial)
        .filter(
            ProjectMaterial.project_id == project_id,
            ProjectMaterial.material_id == material_id,
        )
        .first()
    )
    if not link:
        return False
    db.delete(link)
    db.commit()
    return True


def _project_to_dict(project: Project, db: Session) -> dict:
    links = (
        db.query(ProjectMaterial)
        .filter(ProjectMaterial.project_id == project.id)
        .all()
    )
    materials = []
    for link in links:
        mat = db.query(Material).filter(Material.id == link.material_id).first()
        if mat:
            materials.append({
                "id": mat.id,
                "title": mat.title,
                "category": f"{mat.category_large}/{mat.category_medium}",
                "note": link.note or "",
            })

    return {
        "id": project.id,
        "name": project.name,
        "description": project.description or "",
        "status": project.status,
        "materials": materials,
        "material_count": len(materials),
        "created_at": project.created_at.isoformat() if project.created_at else None,
        "updated_at": project.updated_at.isoformat() if project.updated_at else None,
    }
